ToggleGroupControl: rewrite backdrop animation with framer-motion

ciampo opened the pull request:
<!-- Thanks for contributing to Gutenberg! Please follow the Gutenberg Contributing Guidelines:
https://github.com/WordPress/gutenberg/blob/trunk/CONTRIBUTING.md -->

## What?
<!-- In a few words, what is the PR actually doing? -->

Completely rewrite the `ToggleGroupControl` 's animated backdrop using `framer-motion` and a couple of different approaches to reading size and position of DOM elements.

Fixes https://github.com/WordPress/gutenberg/issues/37506

## Why?
<!-- Why is this PR necessary? What problem is it solving? Reference any existing previous issue(s) or PR(s), but please add a short summary here, too -->

Current implementation is complex, difficult to read, imperative (based off arbitrary timeouts), and [buggy](https://github.com/WordPress/gutenberg/issues/37506)

## How?
<!-- How is your PR addressing the issue at hand? What are the implementation details? -->

Main changes are:

- use `framer-motion` to handle the animation
- Simplify the backdrop positioning and sizing calculations:
    - use `offsetLeft` instead of `getBoundingClientRect().left` in order to get the gap relative to the offset parent (instead of the viewport)
    - calculate all item sizes before hand, so that the animation can run smooth while the user interacts with the component
    - use `useLayoutEffect` in order to make sure that all DOM updates have taken effect before calculating sizes and offsets

## Testing Instructions
<!-- Please include step by step instructions on how to test this PR. -->
<!-- 1. Open a Post or Page. -->
<!-- 2. Insert a Heading Block. -->
<!-- 3. etc. -->

Make sure that the backdrop animation works as expected in Storybook and in the editor (in particular testing for the scenario highlighted in #37506)

## Screenshots or screencast <!-- if applicable -->


https://user-images.githubusercontent.com/1083581/162028461-cf223330-cbe0-43df-b128-94ad0d24f5f3.mp4


https://user-images.githubusercontent.com/1083581/162048005-b7976479-b33c-4128-8127-08a90dee414f.mp4

ciampo committed:
Rewrite `ToggleGroupControlBackdrop` using `framer-motion`

ciampo committed:
Remove unused config

ciampo committed:
Remove unused `containerRef`

ciampo committed:
Remove unused transition styles

ciampo committed:
Remove unused import

ciampo committed:
Use data attribute to select the option wrapper when computing item sizes for backdrop

ciampo committed:
Refine selectedItemIndex check

ciampo committed:
Force `ToggleGroupControlOption` and `ToggleGroupControlOptionIcon` as only allowed components of ToggleGroupControl

ciampo commented on packages/components/src/toggle-group-control/toggle-group-control-option-base/component.tsx:
```diff
@@ -79,7 +79,12 @@ function ToggleGroupControlOptionBase(
+			className={ labelViewClasses }
+			data-active={ isActive }
+			// Necessary for the ToggleGroupControlBackdrop component to render properly
+			data-toggle-group-control-option-wrapper="true"
```
The `data-toggle-group-control-option-wrapper` is used as a way for the backdrop component to always [find the correct `ToggleGroupControl`'s "option wrapper"](https://github.com/WordPress/gutenberg/pull/40107#discussion_r844175416), which is used to compute the correct position and size to apply to the backdrop.

ciampo commented on packages/components/src/toggle-group-control/types.ts:
```diff
@@ -98,7 +104,9 @@ export type ToggleGroupControlProps = Omit<
 	 * React children
 	 */
-	children: ReactNode;
+	children:
```
This change is not necessary, but I thought it could be interesting to enforce only the expected children type via TypeScript

ciampo commented on packages/components/src/toggle-group-control/toggle-group-control/toggle-group-control-backdrop.tsx:
```diff
@@ -1,86 +1,96 @@
-			window.cancelAnimationFrame( animationRequestId );
-		};
-	}, [ canAnimate, containerRef, containerWidth, state, isAdaptiveWidth ] );
+				const toggleGroupOptionWrapper = item.ref.current?.closest(
```
Since the markup of `ToggleGroupControlOptionBase` can change [if the tooltip is enabled or not](https://github.com/WordPress/gutenberg/blob/5b35278128db9d05cf3bc30aa6e50e8b1e640e4a/packages/components/src/toggle-group-control/toggle-group-control-option-base/component.tsx#L32-L41), I decided to [add a data attribute](https://github.com/WordPress/gutenberg/pull/40107#discussion_r844172550) to the option's wrapper and use [`Element.closest()`](https://developer.mozilla.org/en-US/docs/Web/API/Element/closest) to find it each item's wrapper.

ciampo commented on packages/components/src/toggle-group-control/toggle-group-control/toggle-group-control-backdrop.tsx:
```diff
@@ -1,86 +1,96 @@
+	bounce: 0.2,
+	// Transition durations in the config are expressed as a string in milliseconds,
+	// while `framer-motion` needs them as integers in seconds.
+	duration: parseInt( CONFIG.transitionDurationFast, 10 ) / 1000,
```
This felt a bit hacky but I couldn't come up with another way of reusing our config as it's written today

ciampo committed:
CHANGELOG

ciampo committed:
snapshots

ciampo commented on packages/components/src/utils/config-values.js:
```diff
@@ -33,7 +33,6 @@ const TOGGLE_GROUP_CONTROL_PROPS = {
 	toggleGroupControlBackdropBackgroundColor:
 		CONTROL_PROPS.controlSurfaceColor,
 	toggleGroupControlBackdropBorderColor: COLORS.ui.border,
-	toggleGroupControlBackdropBoxShadow: 'transparent',
```
Deleted as it didn't seem like it made any difference

jasmussen commented:
Very nice! The animation for the backdrop is present in the segmented control in this branch, but presumably without the technical challenges that are present in trunk:

![this branch](https://user-images.githubusercontent.com/1204802/162142412-248c7f5e-03df-4ad9-b2f4-77e4668bd23c.gif)

For comparison, here's the trunk version:
![trunk](https://user-images.githubusercontent.com/1204802/162142470-e151cf78-b108-4a53-8d4a-53df6a57a046.gif)

The trunk version feels a bit slower, just a millisecond or so. That makes it feel a bit more alive — still fast enough that if you use the arrowkeys to set the specific size, it feels instant enough, but still slow enough that you the animation is perceptible. Can we slow this framer version down just a teensy bit to better match what's in trunk?

_I realize it's probably hard to discern the speed differential with the reduced framerate of the GIFs above, but nevertheless._

ciampo committed:
Slower transition duration

ciampo commented:
> Can we slow this framer version down just a teensy bit to better match what's in trunk?

I changed it (https://github.com/WordPress/gutenberg/pull/40107/commits/ffde8adc99b41e2530e40577cacceedd5bfbebea) from `160ms` to `200ms`, as that was the slowest transition duration that we currently have [in our config](https://github.com/WordPress/gutenberg/blob/10aa3714ce74a47ed375e6905461a2fbfff280e2/packages/components/src/utils/config-values.js#L80).

If that's still too fast, I'm happy to introduce one more variable — maybe `250ms` ?

jasmussen commented:
I'd love another inspection on this, because it still feels a bit too fast for me. Compare this branch:
![currently](https://user-images.githubusercontent.com/1204802/162191090-58ddd2be-fce5-441f-b3da-58edd12e64ec.gif)

with trunk:

![trunk](https://user-images.githubusercontent.com/1204802/162191092-3298c6ab-f3ad-4d9c-b201-d758c2b404a4.gif)

Trunk still feels a bit more harmonious. The curious thing is, looking at the code, it appears the trunk version was set to 100ms 🤔 

Is there any elastic bounciness going on? It's a bit hard to tell. If yes, I don't think we should have that — ease in/out should be sufficient. Thanks for keeping at this!

ciampo committed:
Remove bounciness

ciampo committed:
Make transition slower

ciampo commented:
I agree that it feels different, it must be a difference between how the animation was implemented previously (raw CSS values) and how `framer-motion` handles it.

> Is there any elastic bounciness going on?

There was indeed some bounciness going on — I went ahead and [removed it](https://github.com/WordPress/gutenberg/pull/40107/commits/8299ee5fd3a9cc5a17721a4bac34e1fe1689adc7), while [increasing the animation duration to `250ms`](https://github.com/WordPress/gutenberg/pull/40107/commits/918c1ffe73dcfc6c7ed4875c07011446b45af85a).

Another thing I noticed is that the previous implementation had [a `100ms` delay that was introduced as a hacky fix to tentatively solve the popover bug](https://github.com/WordPress/gutenberg/blob/trunk/packages/components/src/toggle-group-control/toggle-group-control/toggle-group-control-backdrop.tsx#L52-L56).

Given that the animation is now slower, would you still like me to re-introduce that delay? I'd personally be against adding a delay here, because it would make the overall interaction feel less responsive (e.g. [RAIL model](https://web.dev/rail/)).

jasmussen commented:
The animation actually feels pretty decent now:
![now](https://user-images.githubusercontent.com/1204802/162195399-a1c750c8-518a-4c98-9d08-2c40c1b1c891.gif)

I noticed another thing. The left side looks good:
<img width="296" alt="Screenshot 2022-04-07 at 14 08 49" src="https://user-images.githubusercontent.com/1204802/162195438-b4a0165d-3f2e-4643-9e6a-bf64df4cf3d1.png">

1px border, 2px padding inside. But the right side looks off:

<img width="115" alt="Screenshot 2022-04-07 at 14 08 54" src="https://user-images.githubusercontent.com/1204802/162195517-5f810a79-4d84-44bd-91db-fa73e8879361.png">

ciampo commented:
Mhh, I think that could be caused by subpixel rouding/rendering, and can probably vary depending on the device / screen resolution / browser.

For example, on my external screen, the font size control component in the editor looks fine in Chrome but similar to your screenshot Firefox:

Chrome:

![Screenshot 2022-04-07 at 14 47 26](https://user-images.githubusercontent.com/1083581/162202151-ed39844b-12f4-44e0-aebe-67427592adb0.png)

Firefox:

![Screenshot 2022-04-07 at 14 56 36](https://user-images.githubusercontent.com/1083581/162203777-a9cfaf39-11a7-4dc4-a903-32dee656a397.png)

At the same time, the Storybook example seems fine on both Chrome and Firefox:

Chrome:

![Screenshot 2022-04-07 at 14 59 35](https://user-images.githubusercontent.com/1083581/162204389-23de2695-8846-487c-bb35-8e23015bbd76.png)

Firefox:

![Screenshot 2022-04-07 at 14 57 36](https://user-images.githubusercontent.com/1083581/162203983-4ba04114-2474-46c4-ad2e-2b9f49f571c2.png)

---

Maybe we can look into it in a separate follow-up, if the rest of this PR looks good?

jasmussen commented:
It doesn't appear to be subpixel math on the backdrop itself, I see all whole pixels:

<img width="1460" alt="Screenshot 2022-04-07 at 15 18 55" src="https://user-images.githubusercontent.com/1204802/162208166-c943b67d-62e5-496d-a552-6d32ff96289b.png">

I know it seems like a small detail, but it bugs me 😅 — if we are to address it in a followup as a bugfix, we need to have high confidence that we _can_ fix it. I.e. if we can figure out a solution, it's fine to do separately.

ciampo commented:
Alright, I'll take a look at it next week and see what I can do.

mirka commented on packages/components/src/toggle-group-control/toggle-group-control/toggle-group-control-backdrop.tsx:
```diff
@@ -1,86 +1,96 @@
+	bounce: 0,
+	// Transition durations in the config are expressed as a string in milliseconds,
+	// while `framer-motion` needs them as integers in seconds.
+	duration: parseInt( CONFIG.transitionDurationSlow, 10 ) / 1000,
```
What do you think about just hardcoding the duration value here? I guess because:

- It's a pretty specialized animation anyway
- What if the `CONFIG` values are ever changed from `ms` to `s`
- It seems like overkill to extract a separate `TRANSITION_DURATION_MS` object (Or maybe not? Will it come up a lot in these JS animations?)

No strong opinion though, just throwing it out there.

ciampo replied:
I'm happy to hardcode it, although we may want to reconsider having a more clear, better organised shared "transition duration" settings when we start refactoring our shared configs

mirka commented on packages/components/src/toggle-group-control/toggle-group-control/styles.ts:
```diff
@@ -38,15 +38,13 @@ export const block = css`
 	background: ${ COLORS.gray[ 900 ] };
 	border-radius: ${ CONFIG.controlBorderRadius };
-	box-shadow: ${ CONFIG.toggleGroupControlBackdropBoxShadow };
+	box-shadow: none;
```
Is this rule necessary?

stokesman committed:
Add wrapping element so resize aware hook measures inner dimensions

This makes is easy to calculate a width that will be the same as that
of the option elements and avoid having to measure them. A few minor
style adjustments are included.

stokesman committed:
Add context variables useful for backdrop size/position calculation

And update typings as appropriate.

stokesman committed:
Calculate backdrop size/position instead of measuring

stokesman committed:
Remove unused attributes and keep `radioProps` clean

stokesman committed:
Tweak transition params for trunk parity

stokesman committed:
Snap the shots

stokesman commented:
Doh! 🤦 I had meant to make a separate branch but missed it and pushed directly to this one. I hope it's no trouble @ciampo  and I'm not trying to make a habit of this! I'm going to leave it for now but if you'd like I can drop the commits and force push this back to how you had it.

Before explaining what my changes are about, I want to point out how terribly tricky this seemingly simple animation can be. Marco has done a great job on simplifying it 👏.

So the final issue here was the backdrop (the dark box that slides around to indicate the active item) could end up misaligned from the right side of the toggle control’s border https://github.com/WordPress/gutenberg/pull/40107#issuecomment-1091656895.

To fix that we need to have the backdrop use sub-pixel size and positions as previously provided by `getBoundingClientRect`. The problem with using that function is that its measurement is affected by scaling from containing elements and that's what makes for the unexpected results in the popover for colors (its mount animation scales everything). It's possible to work around that by having the effect that sets the measure hold off until scale is 1:1 but it's a bit more complicated and makes for a late appearance of the backdrop in such contexts.

So what I came up with was to avoid measuring and instead calculate the size and position that the backdrop needs to be. It actually ends up allowing for more simplification but I've just realized the biggest downfall—`isAdaptiveWidth` is not supported!

To support varying sized options, I think measuring will have to be done. Maybe there's a way to apply that strategy only when `isAdaptiveWidth={ true }` without too much complication. Otherwise, if we can't think of a better way to resolve all this maybe we could consider dropping that prop. It seems to be unused in Gutenberg and the component is experimental.

stokesman committed:
Revert "Add wrapping element so resize aware hook measures…"

This reverts commit ed2cd80a42aa43d3f54aaed5868a56820324aeb3.

stokesman committed:
Revert "Snap the shots"

This reverts commit 2ce5baa625b3ac10ae4e175fcfb5d96350cfbdc2.

stokesman committed:
Calculate backdrop size and position with inset from styles

stokesman commented:
I've found a way to support `isAdaptiveWidth` and keep it pretty simple (and also not have to update snapshots). It's not ready for a push just yet and I've got to call a day for now.

ciampo commented:
Thank you for the explanation, @stokesman !

I was also working on a new approach that ideally ditches all calculations altogether by using `framer-motion`'s layout animations — I'll probably open a new branch for this approach anyway

ciampo commented:
New approach with shared layout animations in #40276

stokesman committed:
Restore data-active attribute on option wrapper

stokesman committed:
Measure option widths and pass via props instead of context

stokesman committed:
Make toggle group children expected type in `PostFeaturedImage`

stokesman committed:
Snapshots

stokesman commented:
I've rebased this and updated it so it works with `isAdaptiveWidth` so this should be a viable solution. The key was using `getComputedStyle` for the width measurement because it provides sub-pixel precision without being affected by parent transforms. It can't provide the x position but it's easy to calculate that.

ciampo commented:
Thank you @stokesman for the work here!

I'd say that the preferred approach for the backdrop animation with `framer-motion` would be the one using shared layout animations (attempted in #40276, and not ready yet to be landed), as it avoids most of the explicit size/position calculations.

With that in mind, I'm going to pause working on/reviewing this PR for the time being, and see if we land #40276 in the next days — in case we can't, we can always get back to this PR.

ciampo commented:
Closing in favour of #40276 (which is still currently blocked by https://github.com/framer/motion/issues/1524 )
